### PR TITLE
feat: prompt to uninstall incompatible extensions

### PR DESCRIPTION
**Pull request by ssbarnea**
This will detect presence of known incompatible extensions and prompt user to remove them. Implementation is based on the same feature that redhat.yaml extension has.

### CODE REVIEW

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:25]: This is not a typical js idiom. Instead, just use an `Array.prototype.map()` to convert a list of objects into a different type.

Comment (webknjaz) [reply src/extensionConflicts.ts:25]: You can then chain `Array.prototype.filter()` to drop unnecessary items.

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:40]: It's more straightforward to just use a "for ... of ..." loop
```suggestion
  for (let ext of conflictingExts) {
```

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:53]: Plz use `map()` here.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: src/extensionConflicts.ts
Comment (ssbarnea) [reply src/extensionConflicts.ts:53]: The entire code was copied from yaml extension and I only made the minimum number of changes in order to pass our linting. I do not want to diverge more as it would make harder to backport changes from there.

Comment (webknjaz) [reply src/extensionConflicts.ts:53]: Then it should go in a special `_vendor` location and never be linted.

Comment (ssbarnea) [reply src/extensionConflicts.ts]: I implemented the changes you requested so vendoring is no longer needed.

File: src/extension.ts
Comment (webknjaz) [inline src/extension.ts:20]: Is this some autoformatting change? In Python, it's usually required not to have an empty line right after the function signature.

File: src/extension.ts
Comment (webknjaz) [inline src/extension.ts:15]: Same: plz keep a two-line separator between imports and the module body.

File: src/extension.ts
Comment (webknjaz) [inline src/extension.ts:61]: the indentation here is broken

File: src/extension.ts
Comment (webknjaz) [inline src/extension.ts:76]:
```suggestion
function notifyAboutConflicts(): void {
```

File: src/extension.ts
Comment (webknjaz) [inline src/extension.ts:81]:
```suggestion
}
```

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:65]:
```suggestion
  if (!conflictingExts.length) {
    return;
  }
  if (clickedMsg === uninstallMsg) {
    conflictingExts = [];
  }
  window.showInformationMessage(conflictMsg, uninstallMsg).then(clickedMsg => {
    for (let ext of conflictingExts) {
      commands.executeCommand('workbench.extensions.uninstallExtension', ext.id);
      uninstallingIDs.delete(ext.id);
    }
  });
```

File: src/extension.ts
Comment (ssbarnea) [reply src/extension.ts:20]: Apparently this was not from eslint so I was able to remove it. Still, i think it would be preferable to keep formatting for the linter alone. If there is something it does not do,
we can look into reconfiguring it in a separate change.

File: src/extension.ts
Comment (ssbarnea) [reply src/extension.ts:15]: I fixed it but if you want this to be followed please make a change to eslint config to enforce it. I guess https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/newline-after-import.md should do the trick.

File: src/extension.ts
Comment (ssbarnea) [reply src/extension.ts]: It was fixed.

File: src/extensionConflicts.ts
Comment (cidrblock) [inline src/extensionConflicts.ts:14]: How is this list going to be maintained?

**Review by cidrblock — APPROVED** [review]
Did not pull and test, one question about maintaining the list moving forward, which shouldn't block the merge

File: src/extensionConflicts.ts
Comment (webknjaz) [reply src/extensionConflicts.ts:14]: I wonder if this could be moved to `package.json`

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:3]: There's no `License.txt`. How to handle this correctly?

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:21]: IIRC, the correct check should be
```suggestion
  return typeof obj !== 'undefined' && !uninstallingIDs.has(obj.id);
```

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:56]: This guard expression should probably be the first thing in the function.

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:59]: Why not make the outer function async and avoid the "promise hell" with an `await`?

File: src/extensionConflicts.ts
Comment (ssbarnea) [reply src/extensionConflicts.ts:56]: done

File: src/extensionConflicts.ts
Comment (ssbarnea) [reply src/extensionConflicts.ts:59]: done

File: src/extensionConflicts.ts
Comment (ssbarnea) [reply src/extensionConflicts.ts:21]: done

File: src/extensionConflicts.ts
Comment (ssbarnea) [reply src/extensionConflicts.ts:3]: done

File: src/extensionConflicts.ts
Comment (ssbarnea) [reply src/extensionConflicts.ts:14]: There is no need to move this anywhere as I do not expect us to do much maintenance on it. Probably we will have to modify the list less than once a year. All the other extensions that have a similar method did the same.

File: src/extensionConflicts.ts
Comment (webknjaz) [reply src/extensionConflicts.ts:3]: Did you forget to push?

Comment (ssbarnea) [reply src/extensionConflicts.ts]: yep, done.

File: src/extensionConflicts.ts
Comment (tomaciazek) [reply src/extensionConflicts.ts]: That would be the case for JavaScript, and if you expect the variable to be completely undeclared. Here we obviously do have the variable declared and the TypeScript compiler will prevent using undeclared variables anyway. So, `obj !== undefined` is perfectly fine in TypeScript, and preferable since it's shorter/less complex.

File: src/extensionConflicts.ts
Comment (webknjaz) [reply src/extensionConflicts.ts:14]: Structurally, it's more obvious when it's in some sort of a data file / metadata. Having to update a list rarely is a questionable excuse for suboptimal design decisions. If the metadata is in a wrong place, it's wrong regardless of when it's going to be updated because the problem in question is of a different category.
Anyway, I'm not going to block this PR on this.

File: src/extensionConflicts.ts
Comment (webknjaz) [inline src/extensionConflicts.ts:67]:
```suggestion
  const clickedMsg = await window.showInformationMessage(conflictMsg, uninstallMsg);
  if (clickedMsg !== uninstallMsg) {
    return;
  }
  for (const ext of conflictingExts) {
    commands.executeCommand('workbench.extensions.uninstallExtension', ext.id);
    uninstallingIDs.delete(ext.id);
  }
```

**Comment by webknjaz** [comment]
Plz also fix the commit message / PR title to contain useful info (such as the high-level effect of the change for the user).

**Comment by ssbarnea** [comment]
> Plz also fix the commit message / PR title to contain useful info (such as the high-level effect of the change for the user).

Done, make pr title more in sync with the changelog entry, stating clearly
what this change does.

File: src/extensionConflicts.ts
Comment (ganeshrn) [inline src/extensionConflicts.ts:25]: Replace vscode-yaml with redhat.ansible extension?

**Review by ganeshrn — APPROVED** [review]
LGTM and good merge after minor doc comment is fixed

File: src/extensionConflicts.ts
Comment (ssbarnea) [reply src/extensionConflicts.ts]: Thanks! I fixed that.

**Review by webknjaz — APPROVED** [review]

**Pull request merged**
